raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2394,6 +2394,11 @@ public Optional<OffsetAndEpoch> latestSnapshotId() {
         return log.latestSnapshotId();
     }
 
+    @Override
+    public long logEndOffset() {
+        return log.endOffset().offset;
+    }
```

Comment:
This is not correct in all cases. The leader can have records in the base accumulator that have not been sent to the log. I think you want something along these lines:
```java
    public long logEndOffset() {
        return quorum.maybeLeaderState()
            .map(leader -> leader.accumulator().nextOffset())
            .orElse(log.endOffset().offset);
    }
```

Then we can add this method to BatchAccumulator:
```java
    public long nextOffset() {
        appendLock.lock();
        try {
            return nextOffset;
        finally {
            appendLock.unlock();
        }
    }
```

Reply:
Can we also add tests for this new functionality?

Reply:
> This is not correct in all cases. The leader can have records in the base accumulator that have not been sent to the log.

I don't care about that, though, since I only plan on using this method when the leader becomes active. We will not use it after that.

Reply:
> Can we also add tests for this new functionality?

I will add a test for the `requiredEndOffset` parameter and the `logEndOffset` method.